worldedit-core/src/main/java/com/sk89q/worldedit/internal/expression/Expression.java

```diff
@@ -127,34 +131,44 @@ public double evaluate(double... values) throws EvaluationException {
             ((Variable) invokable).value = values[i];
         }
 
-        Future<Double> result = evalThread.submit(new Callable<Double>() {
-            @Override
-            public Double call() throws Exception {
-                pushInstance();
-                try {
-                    return root.getValue();
-                } finally {
-                    popInstance();
-                }
+        try {
+            if (timeout < 0) {
+                return evaluateRoot();
             }
-        });
+            return evaluateRootTimed(timeout);
+        } catch (ReturnException e) {
+            return e.getValue();
+        } // other evaluation exceptions are thrown out of this method
+    }
+
+    private double evaluateRootTimed(int timeout) throws EvaluationException {
+        Future<Double> result = evalThread.submit(this::evaluateRoot);
         try {
-            return result.get(WorldEdit.getInstance().getConfiguration().calculationTimeout, TimeUnit.MILLISECONDS);
+            return result.get(timeout, TimeUnit.MILLISECONDS);
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
             throw new RuntimeException(e);
+        } catch (TimeoutException e) {
+            result.cancel(true);
+            throw new ExpressionTimeoutException("Calculations exceeded time limit.");
         } catch (ExecutionException e) {
             Throwable cause = e.getCause();
             if (cause instanceof ReturnException) {
-                return ((ReturnException) cause).getValue();
+                throw (ReturnException) cause;
             }
             if (cause instanceof RuntimeException) {
                 throw (RuntimeException) cause;
             }
             throw new RuntimeException(cause);
```

Comment:
Can yo add a case for EvaluationException, so we don't break Exception checking as much by introducing the Future?
```
             if (cause instanceof EvaluationException) {
                throw (EvaluationException) cause;
            }
```

Reply:
will do.